rero_ils/alembic/5f0b086e4b82_patron_role_migration.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019-2022 RERO
+# Copyright (C) 2019-2022 UCLouvain
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Patron role migration."""
+
+from logging import getLogger
+
+from rero_ils.modules.patrons.api import Patron, PatronsIndexer, PatronsSearch
+from rero_ils.modules.users.models import UserRole
+
+# revision identifiers, used by Alembic.
+revision = '5f0b086e4b82'
+down_revision = 'eec683a446e5'
+branch_labels = ()
+depends_on = None
+
+LOGGER = getLogger('alembic')
+indexing_chunck_size = 100
+
+
+def upgrade():
+    """Upgrade database."""
+
+    def get_new_roles(roles):
+        if 'system_librarian' in roles:
+            return [UserRole.FULL_PERMISSIONS]
+        elif 'librarian' in roles:
+            return [
+                UserRole.PROFESSIONAL_READ_ONLY,
+                UserRole.ACQUISITION_MANAGER,
+                UserRole.CATALOG_MANAGER,
+                UserRole.CIRCULATION_MANAGER,
+                UserRole.USER_MANAGER
+            ]
+
+    query = PatronsSearch()\
+        .filter('terms', roles=['librarian', 'system_librarian'])\
+        .source(False)
+    patron_uuids = []
+
+    for hit in query.scan():
+        patron = Patron.get_record(hit.meta.id)
+        original_roles = patron.get('roles')
+        migrated_roles = get_new_roles(original_roles)
+        LOGGER.info(f'* Updating ptrn#{patron.pid} [{patron.formatted_name}]')
+        LOGGER.info(f'\t - Original roles are : [{original_roles}]')
+        LOGGER.info(f'\t - New roles are : [{migrated_roles}]')
+        patron['roles'] = migrated_roles
```

Comment:
What about a user with original roles [`patron`, `librarian`]? They would lose their patron role with this script. We should replace existing roles in the user and not rewrite.

Reply:
It should not happens because (from my memories) we won't have a `Patron` resources with `patron` and `librarian` (it causes some problems for authentication ?) In such case, this `User` must have two related `Patron` resources : One with `patron` role, other with `staff roles`